### PR TITLE
cache data for background session task, allows reconnecting with back…

### PR DESCRIPTION
**Pull request by thuyBox**
…ground session task through associateId

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @thuyBox has signed the CLA. Thanks for the pull request!

File: BoxContentSDK/BoxContentSDK/BOXContentSDKErrors.h [hunk 1]
```diff
@@ -81,7 +81,8 @@ typedef NS_ENUM(NSUInteger, BOXContentSDKURLSessionCacheError) {
     BOXContentSDKURLSessionCacheErrorFileCreateFailed = 50000,
     BOXContentSDKURLSessionCacheErrorNoValidSessionForUserId = 50001,
     BOXContentSDKURLSessionCacheErrorNoValidUserIdOrAssociateId = 50002,
-    BOXContentSDKURLSessionCacheErrorNoValidBackgroundSessionIdOrSessionTaskId = 50003
+    BOXContentSDKURLSessionCacheErrorNoValidBackgroundSessionIdOrSessionTaskId = 50003,
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXContentSDKErrors.h]: *NoValid* -> *Invalid* for these constants.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h [hunk 2]
```diff
@@ -160,4 +173,10 @@
  */
 - (BOOL)deleteCachedInfoForUserId:(NSString *)userId associateId:(NSString *)associateId error:(NSError **)error;
 
+- (NSString *)destinationFilePathGivenBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId;
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h]: Minor: `destinationFilePath*For*BackgroundSessionId`

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h [hunk 3]
```diff
@@ -160,4 +173,10 @@
...
+
+- (NSData *)responseDataGivenBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId;
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h]: Minor: responseData*For*BackgroundSessionId.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h [hunk 4]
```diff
@@ -160,4 +173,10 @@
...
+
+- (BOXURLBackgroundSessionIdAndSessionTaskId *)backgroundSessionIdAndSessionTaskIdGivenUserId:(NSString *)userId associateId:(NSString *)associateId error:(NSError **)error;
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h]: Minor: backgroundSessionIdAndSessionTaskId*For*UserId

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 5]
```diff
@@ -39,8 +38,8 @@ - (id)initWithCacheRootDir:(NSString *)cacheRootDir
             NSString *cacheDir = [cacheRootDir stringByAppendingPathComponent:BOXURLSessionTaskCacheDirectoryName];
             if ([[NSFileManager defaultManager] fileExistsAtPath:cacheDir isDirectory:&isDir] == NO || isDir == NO) {
                 NSError *error = nil;
-                [[NSFileManager defaultManager] createDirectoryAtPath:cacheDir withIntermediateDirectories:YES attributes:nil error:&error];
-                BOXAssert(error != nil, @"Failed to create cacheDir %@ with error %@", cacheDir, error);
+                BOOL success = [[NSFileManager defaultManager] createDirectoryAtPath:cacheDir withIntermediateDirectories:YES attributes:nil error:&error];
+                BOXAssert(success, @"Failed to create cacheDir %@ with error %@", cacheDir, error);
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:90]: What's the benefit of asserting here for developers and/or us? Should we just return nil if it silently fails?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 6]
```diff
@@ -189,6 +187,32 @@ - (BOXURLSessionTaskCachedInfo *)cachedInfoForUserId:(NSString *)userId associat
     return cachedInfo;
 }
 
+- (NSString *)destinationFilePathGivenBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId
+{
+    NSString *filePath = [self filePathForBackgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId type:BOXURLSessionTaskCacheFileTypeDestinationFilePath];
+
+    //decrypt data found at filePath
+    NSData *data = [self unencryptedDataAtFilePath:filePath];
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:336]: What if this is a big data file? Do we want to offload this to a background queue?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 7]
```diff
@@ -189,6 +187,32 @@ - (BOXURLSessionTaskCachedInfo *)cachedInfoForUserId:(NSString *)userId associat
...
+    return [NSKeyedUnarchiver unarchiveObjectWithData:data];
+}
+
+- (NSData *)responseDataGivenBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: Minor: responseData*For*BackgroundSessionId

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 6 above)
Comment (irvingruan) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:336]: I'll say the same about the encryption/decryption methods below.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 8]
```diff
@@ -265,6 +270,31 @@ - (BOOL)cacheBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(
     return success;
 }
 
+- (NSString *)filePathForBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId type:(BOXURLSessionTaskCacheFileType)type
+{
+    NSString *path = [self dirPathOfSessionTaskWithBackgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId];
+    //persist data to sessions/$backgroundSessionId/$sessionTaskId/$fileType
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: Should we add this comment above this method call?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 9]
```diff
@@ -358,6 +388,8 @@ - (BOXURLBackgroundSessionIdAndSessionTaskId *)parseBackgroundSessionIdAndSessio
 
 - (BOXURLBackgroundSessionIdAndSessionTaskId *)backgroundSessionIdAndSessionTaskIdGivenUserId:(NSString *)userId associateId:(NSString *)associateId error:(NSError **)error
 {
+    if (userId == nil || associateId == nil) return nil;
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: Let's wrap this braces and on a different line.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 10]
```diff
@@ -358,6 +388,8 @@ - (BOXURLBackgroundSessionIdAndSessionTaskId *)parseBackgroundSessionIdAndSessio
...
+
     NSString *dir = [self dirPathOfUserSessionTaskGivenUserId:userId associateId:associateId];
     NSError *err;
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: Let's assign err to nil here.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.h [hunk 11]
```diff
@@ -95,44 +97,87 @@ totalBytesExpectedToSend:(int64_t)totalBytesExpectedToSend;
...
+ @return a background download task. Nil if already completed
  */
-- (NSURLSessionDownloadTask *)createBackgroundDownloadTaskWithResumeData:(NSData *)resumeData taskDelegate:(id <BOXURLSessionDownloadTaskDelegate>)taskDelegate;
+- (NSURLSessionDownloadTask *)backgroundDownloadTaskWithRequest:(NSURLRequest *)request taskDelegate:(id <BOXURLSessionDownloadTaskDelegate>)taskDelegate userId:(NSString *)userId associateId:(NSString *)associateId;
 
 /**
- Create a NSURLSessionUploadTask which can be run in background
+ Retrieve a NSURLSessionDownloadTask given a resume data to be run in the background to download file into a destination file path.
+
+ @param resumeData      data to resume download session task from
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionManager.h]: What do you think about using resumableData instead of resumeData?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m [hunk 12]
```diff
@@ -115,17 +164,49 @@ - (NSURLSession *)createBackgroundSession
...
+        for (NSURLSessionUploadTask *uploadTask in uploadTasks) {
+            [self associateBackgroundSessionTask:uploadTask withTaskDelegate:nil];
+        }
+        for (NSURLSessionDownloadTask *downloadTask in downloadTasks) {
+            [self associateBackgroundSessionTask:downloadTask withTaskDelegate:nil];
+        }
+        dispatch_semaphore_signal(sema);
+    }];
+
+    dispatch_semaphore_wait(sema, DISPATCH_TIME_FOREVER);
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m]: The DISPATCH_TIME_FOREVER feels unsafe. Can we set a timeout here and a retry mechanism?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m [hunk 13]
```diff
@@ -167,50 +251,144 @@ - (void)deassociateSessionId:(NSString *)sessionId sessionTaskId:(NSUInteger)ses
...
+{
+    return [self.cacheClient cachedInfoForUserId:userId associateId:associateId error:error];
+}
+
+- (BOOL)cleanUpSessionTaskInfoGivenUserId:(NSString *)userId associateId:(NSString *)associateId error:(NSError **)error
+{
+    if (userId == nil || associateId == nil) {
+        return YES;
+    }
+    NSError *err = nil;
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m]: We got an error and err variable. Let's be specific with what kind of "err" is.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m [hunk 14]
```diff
@@ -167,50 +251,144 @@ - (void)deassociateSessionId:(NSString *)sessionId sessionTaskId:(NSUInteger)ses
...
+{
+    NSError *error = nil;
+    NSURLSessionTask *sessionTask = nil;
+    BOXBackgroundSessionIdAndTask *task = [self existingBackgroundSessionTaskGivenUserId:userId associateId:associateId];
+
+    NSString *backgroundSessionId = nil;
+
+    if (task == nil) {
+        backgroundSessionId = backgroundSessionIdentifier;
+        sessionTask = uploadFromFileURL == nil ? (request != nil ? [self.backgroundSession downloadTaskWithRequest:request] : [self.backgroundSession downloadTaskWithResumeData:resumeDownloadData]) : [self.backgroundSession uploadTaskWithRequest:request fromFile:uploadFromFileURL];
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m]: This is long and hard to parse. Can we split this to multiple lines?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m [hunk 15]
```diff
@@ -167,50 +251,144 @@ - (void)deassociateSessionId:(NSString *)sessionId sessionTaskId:(NSUInteger)ses
...
+    if (task == nil || taskDelegate == nil || userId == nil || associateId == nil) {
+        return;
+    }
+
+    [self associateSessionId:backgroundSessionId sessionTask:task withTaskDelegate:taskDelegate];
+
+    //cache background task if have not
+    NSError *error = nil;
+    BOOL success = [self.cacheClient cacheUserId:userId associateId:associateId backgroundSessionId:backgroundSessionId sessionTaskId:task.taskIdentifier error:&error];
+    BOXAssert(success, @"failed to cache background task %@", error);
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m]: What do you think about bubbling up an error here instead of asserting? I wonder if it's still advantageous to give back the background task even if persisting it to disk somehow failed, but at least there's an error exposed?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m [hunk 16]
```diff
@@ -219,32 +397,86 @@ - (NSURLSessionUploadTask *)createNonBackgroundUploadTaskWithStreamedRequest:(NS
...
+    BOOL success = YES;
+    NSError *err = nil;
+    if (destinationFilePath != nil) {
+        NSFileManager *fileManager = [NSFileManager defaultManager];
+        NSURL *destURL = [[NSURL alloc] initFileURLWithPath:destinationFilePath];
+
+        if (![fileManager fileExistsAtPath:destinationFilePath]) {
+            success = [fileManager moveItemAtURL:temporaryURL toURL:destURL error:&err];
+        } else {
+            success = [fileManager replaceItemAtURL:destURL
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m:535]: Minor: let's align these colons.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m [hunk 17]
```diff
@@ -219,32 +397,86 @@ - (NSURLSessionUploadTask *)createNonBackgroundUploadTaskWithStreamedRequest:(NS
...
 - (void)URLSession:(NSURLSession *)session downloadTask:(NSURLSessionDownloadTask *)downloadTask
 didFinishDownloadingToURL:(NSURL *)location
 {
-    id<BOXURLSessionTaskDelegate> taskDelegate = [self taskDelegateForSessionId:session.configuration.identifier sessionTaskId:downloadTask.taskIdentifier];
-    if ([taskDelegate conformsToProtocol:@protocol(BOXURLSessionDownloadTaskDelegate)]) {
-        id<BOXURLSessionDownloadTaskDelegate> downloadTaskDelegate = (id<BOXURLSessionDownloadTaskDelegate>)taskDelegate;
-
-        if ([downloadTaskDelegate respondsToSelector:@selector(downloadTask:didFinishDownloadingToURL:)]) {
-            [downloadTaskDelegate downloadTask:downloadTask didFinishDownloadingToURL:location];
+    //this method is only called by background download tasks
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m]: Minor: Proper caps and space this comment.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m [hunk 18]
```diff
@@ -219,32 +397,86 @@ - (NSURLSessionUploadTask *)createNonBackgroundUploadTaskWithStreamedRequest:(NS
...
+    if ([downloadTask.response isKindOfClass:[NSHTTPURLResponse class]]) {
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m]: Will .response ever be something different than NSHTTPURLResponse?

File: BoxContentSDK/BoxContentSDK/BOXContentSDKErrors.h (hunk 1 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXContentSDKErrors.h]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h (hunk 2 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h (hunk 3 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h (hunk 4 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 5 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:90]: yeah, assert to terminate in development mode. We will retry to create this dir when building subdir structures later. Do you think we should return nil and allow error as an argument for init so the user is aware of the issue?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 6 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:336]: the data we encrypt/decrypt would be very small since they contain metadata around the session tasks, and they will be called on the thread of BOXAPIOperation.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 7 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 8 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 9 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.h (hunk 11 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionManager.h]: uhm, this name is what Apple uses for their method cancelByProducingResumeData, and downloadTaskWithResumeData, I'm afraid changing it might make ones think it's different from what's used by those methods

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m (hunk 13 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m]: done, renamed the input error as outError instead

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m (hunk 14 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m (hunk 15 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m]: done, exposing error

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m (hunk 16 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m:535]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m (hunk 17 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m (hunk 18 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m]: we don't expect it to be, and if it is, we won't handle it

File: BoxContentSDK/BoxContentSDK/BOXContentSDKConstants.h [hunk 19]
```diff
@@ -444,20 +444,22 @@ extern NSString *const BOXAPIEnterpriseEventTypeItemUnsync;
...
+extern NSString *const BOXURLSessionTaskCacheUserIdAndAssociateId;
 
 typedef NS_ENUM(NSUInteger, BOXURLSessionTaskCacheFileType) {
     BOXURLSessionTaskCacheFileTypeDestinationFilePath = 0,
     BOXURLSessionTaskCacheFileTypeResumeData = 1,
     BOXURLSessionTaskCacheFileTypeResponse = 2,
     BOXURLSessionTaskCacheFileTypeResponseData = 3,
-    BOXURLSessionTaskCacheFileTypeError = 4
+    BOXURLSessionTaskCacheFileTypeError = 4,
+    BOXURLSessionTaskCacheFileTypeUserIdAndAssociateId = 5
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/BOXContentSDKConstants.h:462]: Minor: do we want error to be the last one?

File: BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.h [hunk 20]
```diff
@@ -156,7 +155,12 @@ typedef void (^BOXSessionTaskReplacedBlock)(NSURLSessionTask *oldSessionTask, NS
  */
 @property (nonatomic, readwrite, strong) NSError *error;
 
-@property (nonatomic, readwrite, copy) BOXSessionTaskReplacedBlock sessionTaskReplacedBlock;
+/**
+ * Stores a unique Id to associate with a background download or upload task
+ * Useful to reconnect to the task upon app restarts, or to resume a previously
+ * cancelled background download
+ */
+@property (nonatomic, readwrite, strong) NSString *associateId;
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.h]: Copy for strings.

File: BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m [hunk 21]
```diff
@@ -290,13 +290,42 @@ - (void)start
...
+        //retrieve cached info to finish
+
+        NSString *userId = self.session.user.modelID;
+        NSError *error = nil;
+        BOXURLSessionTaskCachedInfo *cachedInfo = [self.session.urlSessionManager sessionTaskCompletedCachedInfoGivenUserId:userId associateId:self.associateId error:&error];
+
+        if (cachedInfo.response != nil && error == nil) {
+            //get valid cached info for session task, finish this operation
+
+            [self sessionTask:self.sessionTask didFinishWithResponse:cachedInfo.response responseData:cachedInfo.responseData error:cachedInfo.error];
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m:313]: We should check if the delegate implements this method.

File: BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m [hunk 22]
```diff
@@ -519,19 +546,25 @@ - (void)finishURLSessionTaskWithData:(NSData *)data response:(NSURLResponse *)re
...
+        [self finishURLSessionTaskWithData:(responseData != nil ? responseData : self.responseData) response:response error:error];
+    }
 }
 
-- (void)processIntermediateResponse:(NSURLResponse *)response
+- (void)sessionTask:(NSURLSessionTask *)sessionTask processIntermediateResponse:(NSURLResponse *)response
 {
-    [self processResponse:response];
+    @synchronized (self) {
+        [self processResponse:response];
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m]: Should we protect against nil response here?

File: BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m [hunk 23]
```diff
@@ -519,19 +546,25 @@ - (void)finishURLSessionTaskWithData:(NSData *)data response:(NSURLResponse *)re
...
+    }
 }
 
-- (void)processIntermediateData:(NSData *)data
+- (void)sessionTask:(NSURLSessionTask *)sessionTask processIntermediateData:(NSData *)data
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m:564]: Same here.

File: BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m [hunk 24]
```diff
@@ -519,19 +546,25 @@ - (void)finishURLSessionTaskWithData:(NSData *)data response:(NSURLResponse *)re
...
 {
-    [self.responseData appendData:data];
+    @synchronized (self) {
+        [self.responseData appendData:data];
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m]: Same here.

File: BoxContentSDK/BoxContentSDK/Requests/BOXFileUploadNewVersionRequest.m [hunk 25]
```diff
@@ -15,6 +15,7 @@ @interface BOXFileUploadNewVersionRequest ()
 
 @property (nonatomic, readwrite, strong) NSString *localFilePath;
 @property (nonatomic, readwrite, strong) NSString *uploadMultipartCopyFilePath;
+@property (nonatomic, readwrite, strong) NSString *associateId;
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/Requests/BOXFileUploadNewVersionRequest.m]: copy for strings.

File: BoxContentSDK/BoxContentSDK/Requests/BOXFileUploadRequest.m [hunk 26]
```diff
@@ -18,6 +18,7 @@ @interface BOXFileUploadRequest ()
 
 @property (nonatomic, readwrite, strong) NSString *localFilePath;
 @property (nonatomic, readwrite, strong) NSString *uploadMultipartCopyFilePath;
+@property (nonatomic, readwrite, strong) NSString *associateId;
```
Comment (irvingruan) [inline BoxContentSDK/BoxContentSDK/Requests/BOXFileUploadRequest.m]: Copy for strings.

File: BoxContentSDKSampleApp/BoxContentSDKSampleApp/BOXSampleAppDelegate.m [hunk 27]
```diff
@@ -29,8 +29,12 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
     UINavigationController *navController = [[UINavigationController alloc] initWithRootViewController:authenticationController];
     self.window.rootViewController = navController;
     [self.window makeKeyAndVisible];
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
```
Comment (irvingruan) [inline BoxContentSDKSampleApp/BoxContentSDKSampleApp/BOXSampleAppDelegate.m]: Any reason why we're delaying a call to this?

File: BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.h (hunk 20 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.h]: done

File: BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m (hunk 21 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m:313]: it's a required delegate method

File: BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m (hunk 22 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m]: added a FIXME to review if we need a check there. It's a bit more thought needed to make sure we treat nil as valid/invalid based on the type of operations

File: BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m (hunk 23 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m:564]: done

File: BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m (hunk 24 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m]: done

File: BoxContentSDK/BoxContentSDK/Requests/BOXFileUploadNewVersionRequest.m (hunk 25 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/Requests/BOXFileUploadNewVersionRequest.m]: done

File: BoxContentSDK/BoxContentSDK/Requests/BOXFileUploadRequest.m (hunk 26 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/Requests/BOXFileUploadRequest.m]: done

File: BoxContentSDKSampleApp/BoxContentSDKSampleApp/BOXSampleAppDelegate.m (hunk 27 above)
Comment (thuyBox) [reply BoxContentSDKSampleApp/BoxContentSDKSampleApp/BOXSampleAppDelegate.m]: removed

File: BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m (hunk 21 above)
Comment (irvingruan) [reply BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m:313]: Sounds good

**Comment by irvingruan** [comment]
+2
